Implementation of OCT2008 analyzer

michaelongso opened the pull request:
# Background

The dotnet roslyn analyzer supports [CA2016](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2016). However, this does not handle the case when `CancellationToken` is defined in `parent` class.

This work creates `OCT2008` - **Integration test container classes should forward the 'CancellationToken' to methods that take one**


# Results

## Previously
![Screen Shot 2022-09-28 at 3 44 45 pm](https://user-images.githubusercontent.com/15998611/192697544-e22cf389-b37c-450d-80a1-f8c85286f143.png)

## After

![Screen Shot 2022-09-28 at 3 47 37 pm](https://user-images.githubusercontent.com/15998611/192697900-bdac3a57-7053-44a1-958c-ca613a052638.png)

Commit:
Implementation of OCT2008 analyzer

skermajo commented on source/RoslynAnalyzers/Descriptors.cs:
```diff
@@ -74,5 +74,17 @@ static DiagnosticDescriptor GetTestAnalyzerDescriptor(string id, string title, s
+        {
+            return new DiagnosticDescriptor("OCT2008",
+                "Integration test container classes should forward the 'CancellationToken' to methods that take one",
+                "Cancellation Token is defined in integration test base class. Container classes should forward the 'CancellationToken' to methods that take one",
```
Some minor grammar. Suggest:

```plaintext
CancellationToken is defined in the IntegrationTest base class. ...
```

skermajo commented on source/RoslynAnalyzers/Descriptors.cs:
```diff
@@ -74,5 +74,17 @@ static DiagnosticDescriptor GetTestAnalyzerDescriptor(string id, string title, s
+                "Octopus.Testing",
+                DiagnosticSeverity.Info,
+                false, // Testing analysers are disabled by default, these will only be enabled for test projects
+                "Forward the 'CancellationToken' parameter to methods that take one to ensure the operation cancellation notifications gets properly propagated, " 
```
```plaintext
Foward the CancellationToken property to compatible methods to ensure cancellation notifications get properly propagated.
```

skermajo commented on source/RoslynAnalyzers/Testing/Integration/IntegrationTestForwardCancellationTokenToInvocationsAnalyzer.cs:
```diff
@@ -0,0 +1,271 @@
+
+namespace Octopus.RoslynAnalyzers.Testing.Integration
+{
+    /// <summary>
+    /// Specific analyzer for Octopus Integration Test container classes.
+    /// This will show "information" hint when method call can forward CancellationToken, but hasn't.
+    /// For Octopus Integration Test, CancellationToken is defined in the base class.
+    /// The existing dotnet analyzer, CA2016, handle cases where CancellationToken is found on the method context / containing class, but
+    /// not when definition is in parent class. (IE: IntegrationTest)
+    /// This implementation is based (heavily) on CA2016 implementation. (https://github.com/dotnet/roslyn-analyzers/pull/3641)
+    /// </summary>
```
Consider limiting the summary to a short one-sentence description of what the analyzer does, then put all the great context in `<remarks>`.

e.g. summary (feel free to improve): `Hint when integration test method calls could take an inherited CancellationToken`

skermajo commented on source/RoslynAnalyzers/Testing/Integration/IntegrationTestForwardCancellationTokenToInvocationsAnalyzer.cs:
```diff
@@ -0,0 +1,271 @@
+            analysisContext.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        void AnalyzeCompilationStart(CompilationStartAnalysisContext compilationStartAnalysisContext)
```
nit - could probably be static?

skermajo commented on source/RoslynAnalyzers/Testing/Integration/IntegrationTestForwardCancellationTokenToInvocationsAnalyzer.cs:
```diff
@@ -0,0 +1,271 @@
+                        // Verify method has ct overload as last parameter.
+                        InvocationMethodHasCancellationTokenOverload(method, cancellationTokenType, genericTaskType, genericValueTaskType))
+                    {
+                        SyntaxNode? nodeToDiagnose = GetInvocationMethodNameNode(context.Operation.Syntax) ?? context.Operation.Syntax;
```
nit - the compiler was probably hinting to use `var` as preferred here.

michaelongso replied:
I prefer to have explicit type when the variable type is not clear from the context. (IE: `rhs` operation or other prior routine / declaration)

skermajo replied:
Fair enough. The only things to keep in mind though are:

- Consistency. Some variables are `var` while others are specific types. Which is which can be very subjective.
- PrettyBot would insist on `var` in other repos.

michaelongso replied:
Good point. They will be updated to `var` on the `fixer` work stream. Thanks for the pointer!

skermajo replied:
Cheers - no need to go out of your way to change these ones, unless it's easy. It was a bit of a nit pick :)

skermajo commented on source/RoslynAnalyzers/Testing/Integration/IntegrationTestForwardCancellationTokenToInvocationsAnalyzer.cs:
```diff
@@ -0,0 +1,271 @@
+                // Overload is valid if its return type is implicitly convertable
+                var toCompareReturnType = methodToCompareWithAllParameters.ReturnType;
+                var originalReturnType = originalMethodWithAllParameters.ReturnType;
+                if (!toCompareReturnType.IsAssignableTo(originalReturnType))
+                {
+                    // Generic Task-like types are special since awaiting them essentially erases the task-like type.
+                    // If both types are Task-like we will warn if their generic arguments are convertable to each other.
+                    if (IsTaskLikeType(originalReturnType) && IsTaskLikeType(toCompareReturnType) &&
+                        originalReturnType is INamedTypeSymbol originalNamedType &&
+                        toCompareReturnType is INamedTypeSymbol toCompareNamedType &&
+                        TypeArgumentsAreConvertable(originalNamedType, toCompareNamedType))
+                    {
+                        return true;
+                    }
+
+                    return false;
+                }
+
+                return true;
```
I suspect this could be simplified for someone reading the code. In general, prefer 'positive' vs 'negative' if statements. Theoy being, `if (aThing)` is easier to reason about and has less cognitive overhead than `if (!aThing)`.

In this case, I suspect you could do something like the following to reduce nesting?

```csharp
if (toCompareReturnType.IsAssignableTo(...)) {
  return true;
}

if (IsTaskLikeType(...) && IsTaskLikeType(...)) {
  ...
}
```

michaelongso replied:
Simplified as per suggested. The `if (IsTaskLikeType(...))` is updated to `return ...;`

skermajo commented on source/RoslynAnalyzers/Testing/Integration/IntegrationTestForwardCancellationTokenToInvocationsAnalyzer.cs:
```diff
@@ -0,0 +1,271 @@
+                    return false;
+                }
+
+                bool TypeArgumentsAreConvertable(INamedTypeSymbol left, INamedTypeSymbol right)
```
nit `Convertible`

skermajo commented on source/Tests/Testing/Integration/IntegrationTestForwardCancellationTokenToInvocationsAnalyzerFixture.cs:
```diff
@@ -0,0 +1,426 @@
+    public class IntegrationTestForwardCancellationTokenToInvocationsAnalyzerFixture
+    {
+        [TestCase]
+        public async Task No_Diagnostic_ClassesThatHaveNothingToDoWithThisAnalyser()
```
nit - file name uses the American `Analyzer` spelling (as does the project) but this method using the British `Analyser`.

skermajo commented on source/Tests/Testing/Integration/IntegrationTestForwardCancellationTokenToInvocationsAnalyzerFixture.cs:
```diff
@@ -0,0 +1,426 @@
+        [TestCase]
+        public async Task Diagnostic_MethodWithCancellationTokenOverload()
+        {
+            var container = @"
```
nit - inconsistency + this should probably be a `const` anyway. All the strings in above tests are `const`, but everything below is using `var`.

michaelongso replied:
It's all consistent now with `const string container`

skermajo commented on source/RoslynAnalyzers/Testing/Integration/IntegrationTestForwardCancellationTokenToInvocationsAnalyzer.cs:
```diff
@@ -0,0 +1,271 @@
+
+        static (INamedTypeSymbol? cancellationTokenType, INamedTypeSymbol? genericTaskType, INamedTypeSymbol? genericValueTaskType) GetWellKnownTypes(OperationAnalysisContext context)
+        {
+            INamedTypeSymbol? cancellationTokenType = context.Compilation.GetTypeByMetadataName(Constants.Types.SystemThreadingCancellationToken);
+            INamedTypeSymbol? genericTaskType = context.Compilation.GetTypeByMetadataName(Constants.Types.SystemThreadingTasksTask1);
+            INamedTypeSymbol? genericValueTaskType = context.Compilation.GetTypeByMetadataName(Constants.Types.SystemThreadingTasksValueTask1);
```
Per other instances, I suspect the compiler would be hinting to use the preferred `var` here.

michaelongso replied:
Updated. The method definition clearly display the expected return type.

skermajo reviewed (changes requested):
Thanks so much for this 👍 The only real critical fix that would be good to see is the minor grammatical errors in the descriptor. All other suggestions are minor and not blocking, though are good polish things.

Commit:
Fixed grammatical error

Commit:
Consistent in using explicit const string container, instead of var.
Consistent spelling on analyzer in test name.

Commit:
Summary is now concise, with additional details under remarks.
Various null check prior to accessing properties.
Updated to use var where variable type is clear from code context.

skermajo reviewed (approved):
👍 Thanks for the tidy-ups. LGTM.
